11M 8-bit QR code (capacity: 251 bytes) support and ZX Spectrum support

nagydani opened the pull request:
Version 11 QR codes with level M error-correction and checkerboard mask.

Also included:

- Display routine for such codes on the ZX Spectrum.
- ZX Spectrum BASIC program for pre-calculating the layout map.
- Lua program for pre-calculating the RS generator polynomial.

nagydani committed:
Lua tool for generator polynomial coefficients

nagydani committed:
Initial commit

nagydani committed:
Wrong loop counter fixed.

nagydani committed:
Unnecessary stack entry removed.

nagydani committed:
interleaving works properly, ECC still faulty

nagydani committed:
layout correct

nagydani committed:
QR code display finalized

nagydani commented:
**Important note:** at present, not everything works correctly, do not merge. Strings resulting in only the first (50-byte) block containing non-zero bytes decode correctly and are module-for-module identical to those generated with the reference application "qrencode" by Kentaro Fukuchi. However, if anything gets into the second block, there are differences in the ECC part. Trying to find the bug.

leiradel commented on src/qrc11.asm:
```diff
@@ -0,0 +1,333 @@
+
+        ; add ixl for dumb assemblers
+	db $dd
+	add a, l
```
Could this be a macro, i.e. `ADD_IXL`? I think it would make the code easier to read.

nagydani replied:
Sure. One sec.

nagydani committed:
Macros for some undocumented Z80 instructions dealing with IXH and IXL for
assemblers that cannot handle such opcodes.

nagydani commented:
It seems that the last four ECC fields are computed on the wrong data. These are encodings by <code>qrencode</code> (Q47 and Q48) and my program (S47 and S48) of 251-byte strings of zeroes with only one capital X at 47th and 48th positions, respectively, which is precisely the boundary between the first and the second block. Q47 is identical to S47, but Q48 and S48 differ.
![Q47](https://user-images.githubusercontent.com/374934/97106203-7eb11b00-16c0-11eb-8994-2f9686dc5a84.png)
![Q48](https://user-images.githubusercontent.com/374934/97106204-7fe24800-16c0-11eb-8905-485cdb52d919.png)
![S47](https://user-images.githubusercontent.com/374934/97106207-807ade80-16c0-11eb-911e-4ac4e189e139.png)
![S48](https://user-images.githubusercontent.com/374934/97106208-807ade80-16c0-11eb-9c1a-1bc84dd59e68.png)

leiradel commented:
I really don't know enough about QRCode to see where the problem is. What I did during the development of `qrc1.asm` was using [this](https://github.com/floooh/chips/blob/master/chips/z80.h) Z80 emulator to run the encoder and print information about each step of the process.

nagydani committed:
Check for empty string; it is not allowed.

nagydani committed:
Interleaving fixed.

nagydani commented:
I have found the bug: the distance between data blocks and ECC blocks is not the same, so interleaving was buggy (used the same routine). Now I have fixed it. Also fixed the ZX Spectrum API so that it rejects the empty string along with excessively long strings.
Anyhow, the code is now working and ready to be merged.

nagydani committed:
User interface

leiradel commented on src/qrc11.asm:
```diff
@@ -0,0 +1,359 @@
+    ; Calculate the message ECC.
+    ; ------------------------------------------------------------------------
+
+    ; Copy each block of the original encoded message to the target buffer,
```
I think this comment doesn't hold anymore, `qrc11_ecc` will save and restore the message block.

leiradel commented on src/qrc11.asm:
```diff
@@ -0,0 +1,359 @@
+    ; Display QR code with checkerboard mask.
+    ; ------------------------------------------------------------------------
+
+    ld hl, qrc11_map
```
Ideally this would be in a different file. `qrc1.asm` is just the platform-agnostic QRCode Version 1 encoder, it's the `zx81.asm` file that has the ZX81 platform specific code that reads a string from the keyboard, encodes it, and prints the resulting QRCode on the screen.

Could you leave just the encoder in `qrc11.asm` and move all the ZX Spectrum specific files into a `ZXSpectrum` folder? Thanks.

leiradel replied:
Some clarification: this is to make it easier for people to use the encoder in other platforms based on the Z80.

leiradel replied:
I see that there's already a `spectrum` folder, the name is fine, no need to rename to `ZXSpectrum`.

Also, `qrc11_module` is defined in a different file, so there are calls from `qrc11.asm` into `spectrum.asm` making the two highly coupled. Maybe your intention was that the platform specific file provide a `qrc11_module` routine? I still prefer that `qrc11.asm` has just the encoder, and that `spectrum.asm` uses it to encode the string, and then uses the platform specific code to put it onto the screen.

leiradel reviewed (approved):
Actually I think you're right, it's the specific QR Code implementation that knows how to print itself, it should only call subroutines defined elsewhere to set pixels and move the cursor around. I'll change `qrc1.asm` and the ZX81 printer to do that.

leiradel commented:
Thanks again for your PR. I've changed your code to work similarly to `qrc1.asm`, and got rid of the large binary file with the bit indices to plot the QR Code.

Please tell me how you'd like to be thanked in my repo, I'll add a link to your website, GitHub profile, or any other place that you prefer, acknowledging you as the original author of the QR Code version 11 code.
